### PR TITLE
Fix incorrect handling of 'exp', 'nbf' and 'iat' claim values

### PR DESCRIPTION
**Pull request by ChenYan98**

### CODE REVIEW

**Comment by brentschmaltz** [comment]
@ChenYan98 We need to have some tests.

**Comment by RussKie** [comment]
Can you verify you can decrypt the token provided here and the values are the same
https://github.com/dotnet/corefx/issues/11082#issuecomment-242279236
Thank you

**Comment by ChenYan98** [comment]
@RussKie , @brentschmaltz Add two test cases. The token in one test case is from https://github.com/dotnet/corefx/issues/11082#issuecomment-242279236. Please take a look.

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs [hunk 1]
```diff
@@ -463,6 +464,99 @@ public void InboundOutboundClaimTypeMapping()
             RunClaimMappingVariation(jwt, handler, validationParameters, expectedClaims: expectedClaims, identityName: null);
         }
 
+#pragma warning disable CS3016 // Arrays as attribute arguments is not CLS-compliant
+        [Theory, MemberData("ReadJwtTokenTheoryData")]
+#pragma warning restore CS3016 // Arrays as attribute arguments is not CLS-compliant
+        public void ReadJwtToken(JwtTheoryData theoryData)
+        {
+            TestUtilities.WriteHeader($"{this}.ReadJwtToken", theoryData);
```
Comment (brentschmaltz) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: Use the new method: TestUtilities.WriteHeader that returns a CompareContext

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs [hunk 2]
```diff
@@ -463,6 +464,99 @@ public void InboundOutboundClaimTypeMapping()
...
+            var context = new CompareContext($"{this}.ReadJwtToken, {theoryData.TestId}");
+            try
+            {
+                var principal = theoryData.TokenHandler.ValidateToken(theoryData.Token, theoryData.ValidationParameters, out SecurityToken validToken);
+
+                var jwtToken1 = validToken as JwtSecurityToken;
+                var jwtToken2 = theoryData.SecurityToken as JwtSecurityToken;
+
+                string time1 = Convert.ToString(jwtToken1.Payload[JwtRegisteredClaimNames.Nbf], CultureInfo.InvariantCulture);
```
Comment (brentschmaltz) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: Why do we need to convert to string to compare?
If the IdentityComparer will not compare two doubles, we need to fix that.

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs [hunk 3]
```diff
@@ -463,6 +464,99 @@ public void InboundOutboundClaimTypeMapping()
...
+                return new TheoryData<JwtTheoryData>
+                {
+                    new JwtTheoryData
+                    {
+                        First = true,
+                        SecurityToken = jwtToken1,
+                        TestId = "CreateTokenFromTokenHandler",
+                        Token = handler.WriteToken(jwtToken1),
+                        ValidationParameters = new TokenValidationParameters
+                            {
```
Comment (brentschmaltz) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: please fix indenting...

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs [hunk 4]
```diff
@@ -463,6 +464,99 @@ public void InboundOutboundClaimTypeMapping()
...
+                            }
+                    },
+
+                    new JwtTheoryData
+                    {
+                        SecurityToken = jwtToken2,
+                        TestId = "TokenFromCustomerReport",
+                        Token = token2,
+                        ValidationParameters = new TokenValidationParameters
+                            {
```
Comment (brentschmaltz) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: please fix indenting...

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs [hunk 5]
```diff
@@ -463,6 +464,99 @@ public void InboundOutboundClaimTypeMapping()
...
+        public static TheoryData<JwtTheoryData> ReadJwtTokenTheoryData
+        {
+            get
+            {
+                TimeSpan timeSpan = DateTime.UtcNow.ToUniversalTime().AddMinutes(10) - EpochTime.UnixEpoch;
+                JwtSecurityTokenHandler handler = new JwtSecurityTokenHandler();
+
+                JwtSecurityToken jwtToken1 = handler.CreateToken(new SecurityTokenDescriptor()) as JwtSecurityToken;
+                jwtToken1.Payload[JwtRegisteredClaimNames.Nbf] = 1471924298.67746;
+                jwtToken1.Payload[JwtRegisteredClaimNames.Iat] = jwtToken1.Payload[JwtRegisteredClaimNames.Nbf];
```
Comment (brentschmaltz) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: why not just assign the value as above '= 1471924298.67746'

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs [hunk 6]
```diff
@@ -463,6 +464,99 @@ public void InboundOutboundClaimTypeMapping()
...
+                jwtToken1.Payload[JwtRegisteredClaimNames.Exp] = timeSpan.TotalSeconds;
```
Comment (brentschmaltz) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: Why do we need this complicated calculation?

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs [hunk 7]
```diff
@@ -463,6 +464,99 @@ public void InboundOutboundClaimTypeMapping()
             RunClaimMappingVariation(jwt, handler, validationParameters, expectedClaims: expectedClaims, identityName: null);
         }
 
+#pragma warning disable CS3016 // Arrays as attribute arguments is not CLS-compliant
+        [Theory, MemberData("ReadJwtTokenTheoryData")]
+#pragma warning restore CS3016 // Arrays as attribute arguments is not CLS-compliant
+        public void ReadJwtToken(JwtTheoryData theoryData)
```
Comment (brentschmaltz) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: This method name should reflect the test. I suggest: ReadTimesExpressedAsDoubles

Comment (ChenYan98) [reply test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: fixed

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs (hunk 1 above)
Comment (ChenYan98) [reply test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: fixed

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs (hunk 2 above)
Comment (ChenYan98) [reply test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: Assign the value as "1471924298.67746", then compare.

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs (hunk 5 above)
Comment (ChenYan98) [reply test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: Agree. Same as above.

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs (hunk 6 above)
Comment (ChenYan98) [reply test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: I try to assign a valid expire datetime to  'Exp'.

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs (hunk 3 above)
Comment (ChenYan98) [reply test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: fixed

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs (hunk 4 above)
Comment (ChenYan98) [reply test/System.IdentityModel.Tokens.Jwt.Tests/JwtSecurityTokenHandlerTests.cs]: fixed

**Comment by brentschmaltz** [comment]
:shipit:

**Comment by RussKie** [comment]
Thank you
When should we expect to have it released?

**Pull request merged**